### ISSUE TITLE
Water2 fragment shader does not support Z-up WCS

### ISSUE THREAD

**Issue by viktorpts** [mrdoob/three.js#17390]
##### Description of the problem
The calculation which blends reflection and refraction fragment color assumes a default Y-up coordinate system to determine the blend amount. As a result, the transition between reflectance and refractance in a Z-up scene (or if the water surface doesn't lie in the xz-plane) happens while the camera is orbited around the scene, instead of while being tilted up and down. I was able to adjust the shader by trial and error, adding the following to line 339 of `examples/jsm/objects/Water2.js`:

```glsl
vec3 adjustedEye = vec3(toEye.x, toEye.z, toEye.y);
float theta = max( dot( adjustedEye, normal ), 0.0 );
```

Which simply swaps the z and y coordinates for this calculation only. It would be nice if the shader works for any camera-up and plane orientation by taking into account the normal of the surface to which it is applied, however my lack of GLSL knowledge prevents me from coming up with a solution.

##### Three.js version
- [x] r106

##### Browser
- [x] All of them

##### OS
- [x] All of them

**Comment by Mugen87** [comment]
In general, we do not ensure that the example code like shaders work in all possible camera configurations. In most cases, it's up to the application code to make the necessary adjustments.

What are the reasons for using `z-up` anyway? I think in many cases you can make your life easier by stick to the default `y-up` convention.

**Comment by EliasHasle** [comment]
I stumbled upon this line while investigating my own Water2 problem in #17251. Wonder if it is relevant here, as it explicitly sets the up direction of the virtual camera to y-up regardless of the up direction of the main camera: https://github.com/mrdoob/three.js/blob/1ee9309171704d23fac425ad48c4eca1aa5b1194/examples/js/objects/Reflector.js#L104

**Comment by viktorpts** [comment]
> In general, we do not ensure that the example code like shaders work in all possible camera configurations. In most cases, it's up to the application code to make the necessary adjustments.

The discussion in this issue tracker is quite the resource for problem-solving (for me, at least), so I figured my solution can be used as reference by anyone who encounters a similar problem and searches for it online or in this repo, even if the issue is closed.

> What are the reasons for using `z-up` anyway? I think in many cases you can make your life easier by stick to the default `y-up` convention.

It's not just the world orientation, if you rotate the surface with the shader on the x- or z-axis, it wont work correctly. In the concrete case I'm dealing with an app that has a 2D world rendered in 3D, so the third coordinate is redundant and I figured I'd save some mental effort by keeping everything in (x,y). It's also the convention for many engineering and architectural applications which I'm used to, so this also helps.

> I stumbled upon this line while investigating my own Water2 problem in #17251. Wonder if it is relevant here, as it explicitly sets the up direction of the virtual camera to y-up regardless of the up direction of the main camera (there is no matching line in Refractor.js, BTW):

This has no effect, as it is overwritten further in the code - everything is copied from the scene and camera with a transformation applied, using the reflecting surface's normal as reference, so the WCS-up is irrelevant - I even rendered the output to a separate canvas to make sure. It's only the mixing shader that assumes  y-distance == distance from the surface.

**Comment by FishOrBear** [comment]
I also use z-up

**Comment by msdog** [comment]
I also use z-up, is there any way to solve it? Or are there any other alternative solutions

**Comment by EliasHasle** [comment]
> I also use z-up, is there any way to solve it?

Yes. For most stuff, you can just keep your z-up scene in a rotated container. I think you need to rotate it by `-Math.PI/2` around the X axis to get a Z-up system within the Y-up system.

**Comment by msdog** [comment]
I found a project that can use Z-up, but the version is not compatible with the latest Threejs
 https://github.com/jbouny/ocean
 https://jbouny.github.io/ocean/demo_test/

**Comment by Mugen87** [comment]
Closing. Supporting all possible up axis permutations in the project is simply out of scope. The issue is not only about the effects also other components like `PointerLockControls` or `FirstPersonControls` assume +Y up. So please use  +Y up whenever possible for best compatibility.
